043-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,112 @@
+for a Kafka cluster that is using KRaft rather than ZooKeeper. This includes a KRaft 
+cluster with combined nodes and one that has separate controller and broker nodes.
+
+## Current situation
+
+These are the current probes used in Strimzi when ZooKeeper is present:
+
+|Component| Liveness | Readiness |
+|---|---|---|
+| ZooKeeper | "imok" from 127.0.0.1:12181 | "imok" from 127.0.0.1:12181 |
+| ZK-based broker | if (ready) { listening on 9091 } else { have ZK session } | if (ready) { return true } |
+
+The "ready" check in both cases is based on the existence of the file `../kafka-ready` which is 
+created by the KafkaAgent when the broker state metric has value 3 (RUNNING).
+
+## Background
+
+This section contains some background information about the ports and metrics we could use for checks.
+
+### Ports
+In KRaft mode there are two different ports that the node could be listening on.
+If it has a controller process it should be listening on the port defined as the `controller.listener.names` which in Strimzi is currently hard-coded to 9090.
+If it has a broker process it should be listening on the replication port which is hard-coded to 9091.
+If the node is running as a combined mode, it should be listening on both ports
+
+### BrokerState
+The expected [BrokerState](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/BrokerState.java) transitions are:
+
+NOT_RUNNING (0) -> STARTING (1) -> RECOVERY (2) -> RUNNING (3) -> PENDING_CONTROLLED_SHUTDOWN (6) -> SHUTTING_DOWN (7)
+
+The other possible value is 127 which represents UNKNOWN.
+
+### Current state metric and quorum
+
+Some useful things to be aware of with the current-state metric and how the controller quorum affects the broker nodes:
+
+* The [current-state](https://kafka.apache.org/documentation/#kraft_quorum_monitoring) metric indicates whether a controller is currently a leader or follower.
+If this metric shows one of these two states, this means the leader election has happened and a controller quorum has been successfully formed.
+* The current-state metric is not accessible to the existing KafkaAgent because it is not a Yammer metric.
+* The broker nodes will not move to a RUNNING state until the quorum leader election has happened.
+
+## Proposal
+
+There are three different types of liveness/readiness probe that we need to consider, first for a combined node, 
+then for a broker and then for a controller.
+
+### Combined mode
+
+The proposed probes are:
+
+* Liveness: node is listening on 9090 and 9091
+* Readiness: node is listening on 9090 and 9091
+
+### Controller only mode
+
+The proposed probes are:
+
+* Liveness: node is listening on 9090
+* Readiness: node is listening on 9090
+
+### Broker only mode
+
+The proposed probes are:
+
+* Liveness: the BrokerState metric < 3 || listening on 9091
+* Readiness: the BrokerState metric >=3 && != 127
```

Comment:
Adding comment from @scholzj that got lost in the file rename:

[scholzj]
Why is this not part of the combined node readiness as well? For example in an similar form to how it works today with ZooKeeper when the Readiness probe essentially changes after reaching some point?

[katheris]
I've explained in the rejected alternatives section why the BrokerState metric check isn't included in the combined node readiness check. Does that answer your question sufficiently or is there something still unclear there?

Reply:
I get that it cannot be relied upon exclusively. But can't it be used in the second phase, after the quorum is formed?

I mean, long term, we plan to support all 3 modes so one would expect that the probes should be in the long term good for all modes. So this seems a bit weird:
* If checking for listening on 9091 is fine in combined mode, should it be good enough in broker only mode?

Or the other way around
* If checking the port 9091 is not good enough for broker-only mode, why is it good enough for combined mode?

This makes me a bit nervous TBH as it seems there is some gap. The rejected alternatives explain some of the issue. But I don't think it completely answers my question to about switching what the probe does after the broker starts working.

Maybe it would be also good to see if we have some suggestion for Kafka how to improve this / or get some suggestion from Kafka how to do this, since this is no way unique issue for Strimzi. That might be beyond the scope of this proposal - but it is worth to at least cover that as a prerequisite to do something better.

Reply:
I think in an ideal world we would check the BrokerState metric once the quorum is formed. So in combined we would use the current-state metric to determine whether the BrokerState should be checked or not. Then in broker only mode we can just always check BrokerState and the pod won't become ready until the quorum is formed.

However, the problem we have is in combined mode we can't get to the current-state metric via the Kafka Agent since it isn't a Yammer metric. So we have two options:
1. Ignore the BrokerState metric for now in combined mode and see if we hit problems where the brokers get stuck in the STARTING or RECOVERY phase
2. Change the way the Kafka Agent works, or find some other workaround to get hold of the current-state metric

Reply:
I thought that the Yammer metrics issue is about the controller states but not the broker states.

Reply:
Yes, but we need a mechanism to know when to check the broker state. We could e.g. have the check be `(brokerState >= 3 && brokerState != 127) || listening on 9091` but is that actually any better than just checking if it's listening on 9091. Because you would never know if the first part ever succeeds.

Thinking about it this way, I feel like the broker **liveness** check should just be listening on 9091 shouldn't it? What's the benefit of having the second part of the check? It doesn't actually enforce that we ever reach that state

Reply:
Yeah,  ut as I said ... if `listening on 9091` is good enough why not use it in the broker. So one way or another, this shows some limitation which needs to be addressed going long-term. I think the operator will anyway need to know if the quorum is formed or not ... so we will anyway need to get this information at some point I guess.